roxie/ccd/ccdqueue.cpp

```diff
@@ -636,6 +639,29 @@ class CRoxieQueryPacket : public CRoxieQueryPacketBase, implements IRoxieQueryPa
         }
         return createSerializedRoxiePacket(mb);
     }
+
+    virtual void noteTimeSent() override
+    {
+        acknowledged = false;
+        timeFirstSent = msTick();
+        if (!timeFirstSent)
+            timeFirstSent = 1;
+    }
+
+    virtual void setAcknowledged() override
+    {
+        acknowledged = true;
+    }
+
+    virtual bool isAcknowledged() const override
+    {
+        return acknowledged;
+    }
+
+    virtual bool resendNeeded(unsigned timeout) const override
+    {
+        return timeFirstSent && !acknowledged && msTick()-timeFirstSent > timeout;
```

Comment:
Would it be be better to pass in an expiry time, so that msTick() is called once in the caller, rather than potentially multiple times.  A trade off in likelihoods.  Are there likely to be a reasonable number of packets that have been sent but timeout has not happened?

It could be optimized by passing in a timeout class that called and cached msTick() on demand if required.  I doubt that is worth it though.